consensus/istanbul/types.go

```diff
@@ -343,6 +411,65 @@ type Message struct {
 	Msg       []byte
 	Address   common.Address // The sender address
 	Signature []byte         // Signature of the Message using the private key associated with the "Address" field
+
+	// The below fields are not serializable since they are private, they are
+	// set when calling Message.FromPayload, only one will be set in any
+	// instance, which is set depends on the Message.Code.
+	committedSubject    *CommittedSubject
+	prePrepare          *Preprepare
+	prepare             *Subject
+	roundChange         *RoundChange
+	queryEnode          *QueryEnodeData
+	forwardMessage      *ForwardMessage
+	enodeCertificate    *EnodeCertificate
+	versionCertificates []*VersionCertificate
+	valEnodeShareData   *ValEnodesShareData
+}
+
+// NewMessage constructs a message with the innerMessage instance and the
+// sender address, it will panic if innerMessage is not a recognised inner
+// message type.
+func NewMessage(innerMessage interface{}, sender common.Address) *Message {
+	message := &Message{
+		Address: sender,
+	}
+	switch t := innerMessage.(type) {
+	case *Preprepare:
+		message.Code = MsgPreprepare
+		message.prePrepare = t
+	case *Subject:
+		message.Code = MsgPrepare
+		message.prepare = t
+	case *CommittedSubject:
+		message.Code = MsgCommit
+		message.committedSubject = t
+	case *RoundChange:
+		message.Code = MsgRoundChange
+		message.roundChange = t
+	case *QueryEnodeData:
+		message.Code = QueryEnodeMsg
+		message.queryEnode = t
+	case *ForwardMessage:
+		message.Code = FwdMsg
+		message.forwardMessage = t
+	case *EnodeCertificate:
+		message.Code = EnodeCertificateMsg
+		message.enodeCertificate = t
+	case []*VersionCertificate:
+		message.Code = VersionCertificatesMsg
+		message.versionCertificates = t
+	case *ValEnodesShareData:
+		message.Code = ValEnodesShareMsg
+		message.valEnodeShareData = t
+	default:
+		panic(fmt.Sprintf("attempt to construct Message unrecognised inner message type %T", t))
+	}
+	bytes, err := rlp.EncodeToBytes(innerMessage)
+	if err != nil {
+		panic(fmt.Sprintf("attempt to serialise inner messgae of type %T failed", innerMessage))
```

Comment:
Hey @trianglesphere, thanks for taking a look at this. Just to clarify, the approach I'm taking here is to panic when we encounter a code error. If the function is called with the wrong type or nil it will panic. This seems to be the approach that we use throughout the codebase, for example.

https://github.com/celo-org/celo-blockchain/blob/4e42e7e8d683929b1317859d86431d5b3d3037c6/consensus/istanbul/backend/backend.go#L427-L429
https://github.com/celo-org/celo-blockchain/blob/4e42e7e8d683929b1317859d86431d5b3d3037c6/consensus/istanbul/backend/backend.go#L431-L433
https://github.com/celo-org/celo-blockchain/blob/4e42e7e8d683929b1317859d86431d5b3d3037c6/consensus/istanbul/backend/backend.go#L686-L688

In all of these cases and many more, the approach we take is to panic if a nil parameter is passed to a function (coding error). So I'd like to understand what it is about this particular panic that makes you uncomfortable but not all the others that we have in the codebase?

Reply:
Sorry, wrong line: I meant it for `panic(fmt.Sprintf("attempt to construct Message unrecognised inner message type %T", t))`. For that, it's easy enough to have a specialization of the new message function for each type which would completely eliminate that panic. That would help us in the case that an incorrect type got passed into the function (which given we take `interface{}` is a logic error, but also something that we can eliminate at compile time).

For this panic, I think it's probably fine. Removing it requires changing rlp.Encode to not error, and my understanding of what would cause rlp to error is that it has an internal logic error or you're trying to encode an incorrect type. I think it'd be hard to accidentally pass a bad type to be RLP encoded here.

Reply:
So just to clarify, you are saying that you are happy with this panic, but only in the case that we create a set of typesafe constructors?

So in this case:
```
NewQueryEnodeMessage(en *queryEnode, address common.Address)
```

If rlp decode fails its because *queryEnode is not rlp encodable, and in that case you are ok with the panicking.

But in this case:
```
NewMessage(inner interface{}, address common.Address)
```

If rlp decode fails its because whatever type `inner` is,  cannot be rlp encoded, and in that case you would be against panicking?

Reply:
In the second case, I don't like the panic if we hit the default case of the switch on the type. I'm actually a little split on panicking if the rlp encode fails, but that is less risky than panicking if we pass the wrong type to `interface {}`.

Reply:
So as discussed I'm going to make this change on this PR.